mat/diagonal.go

```diff
@@ -85,41 +87,150 @@ func (d *DiagDense) TTri() Triangular {
 	return TransposeTri{d}
 }
 
+// TBand performs an implicit transpose by returning the receiver inside a
+// TransposeBand.
 func (d *DiagDense) TBand() Banded {
 	return TransposeBand{d}
 }
 
+// Bandwidth returns the upper and lower bandwidths of the matrix.
+// These values are always zero for diagonal matrices.
 func (d *DiagDense) Bandwidth() (kl, ku int) {
 	return 0, 0
 }
 
 // Symmetric implements the Symmetric interface.
 func (d *DiagDense) Symmetric() int {
-	return len(d.data)
+	return d.n
 }
 
 // Triangle implements the Triangular interface.
 func (d *DiagDense) Triangle() (int, TriKind) {
-	return len(d.data), Upper
+	return d.n, Upper
 }
 
+// Reset zeros the length of the matrix so that it can be reused as the
+// receiver of a dimensionally restricted operation.
+//
+// See the Reseter interface for more information.
+func (d *DiagDense) Reset() {
+	// No change of Inc or n to 0 may be
+	// made unless both are set to 0.
+	d.mat.Inc = 0
+	d.n = 0
+	d.mat.Data = d.mat.Data[:0]
+}
+
+// DiagOf extracts the diagonal of m into the receiver. If m can return
```

Comment:
IMO, this is wrong. This should be `DiagFrom`, and copy the result into the receiver. This works the same for any `Matrix`. Then, on `Dense`, `Symmetric`, etc. there should be a `func (d *Dense) DiagView() Diagonal` method that returns a view of the diagonal.

Reply:
No worries. I'll change this.